crypto/bio/bio.c

```diff
@@ -68,34 +68,110 @@
+                              size_t len, int argi, long argl, int bio_ret,
+                              size_t *processed) {
+  assert(bio != NULL);
+  assert(bio->callback != NULL);
+
+  /* Strip off any BIO_CB_RETURN flag */
+  int bareoper = oper & ~BIO_CB_RETURN;
+
+  if (bareoper == BIO_CB_READ || bareoper == BIO_CB_WRITE
+    || bareoper == BIO_CB_GETS) {
+    /* In this case |len| is set, and should be used instead of |argi| */
+    if (len > INT_MAX) {
+      return -1;
     }
-    // Pass the original BIO's return value to the callback. If the callback
-    // is successful return processed from the callback, if the callback is
-    // not successful return the callback's return value.
-    long callback_ret = bio->callback_ex(bio, oper, buf, len, 0, 0L, ret, &processed);
-    if (callback_ret <= INT_MAX && callback_ret >= INT_MIN) {
-      ret = (int)callback_ret;
-      if (ret > 0) {
-        // BIO will only read int |len| bytes so this is a safe cast
-        ret = (int)processed;
-      }
+
+    argi = (int)len;
+  }
+
+  if (bio_ret && (oper & BIO_CB_RETURN) && bareoper != BIO_CB_CTRL) {
+    if (*processed > INT_MAX) {
+      return -1;
+    }
+
+    bio_ret = *processed;
+  }
+
+
+  long ret = bio->callback(bio, oper, argp, argi, argl, bio_ret);
+
+  if (ret >= 0 && (oper & BIO_CB_RETURN) && bareoper != BIO_CB_CTRL) {
+    *processed = (size_t)ret;
+    ret = 1;
+  }
+
+  return ret;
+}
+
+//   |get_callback| returns the appropriate callback function for a given |BIO|, preferring
+//   the extended interface |callback_ex| over the legacy interface.
+//
+//   When only the legacy callback is available, it is wrapped in the extended format
+//   via |callback_fn_wrap_ex| to provide a consistent interface. The extended callback
+//   provides additional parameters for length and bytes processed tracking.
+//
+//   Returns the |callback_ex| function if available, a wrapped legacy callback if only
+//   |callback| is set, or NULL if no callbacks are set.
+static BIO_callback_fn_ex get_callback(BIO *bio) {
+  if (bio == NULL) {
+    return NULL;
+  }
+
+  if (bio->callback_ex != NULL) {
+    return bio->callback_ex;
+  }
+  if (bio->callback != NULL) {
+    // Wrap old-style callback in extended format
+    return callback_fn_wrap_ex;
+  }
+  return NULL;
+}
+
+// Helper function to handle return values from |BIO_read|, |BIO_write|,
+// |BIO_gets|, and |BIO_puts| operations.
+static int handle_callback_return(BIO *bio, int oper, const void *buf,
+int len, int ret) {
+
+  size_t processed = 0;
+
+  if (ret > 0) {
+    if (oper == BIO_CB_READ || oper == BIO_CB_GETS) {
+      bio->num_read += ret;
+    } else if (oper == BIO_CB_WRITE || oper == BIO_CB_PUTS) {
+      bio->num_write += ret;
+    }
+    processed = ret;
+    ret = 1;
+  }
+
+  BIO_callback_fn_ex cb = get_callback(bio);
+  if (cb != NULL) {
+    ret = cb(bio, oper | BIO_CB_RETURN, buf, len, 0, 0L, ret, &processed);
```

Comment:
* `ret` is of type `int, but `cb` returns a `long`.
* I think this should be a separate variable and the range check pulled inside the block:
```
  if (cb != NULL) {
    long callback_ret = cb(bio, oper | BIO_CB_RETURN, buf, len, 0, 0L, ret, &processed);
    if (callback_ret > INT_MAX || callback_ret < INT_MIN) {
      return -1;
    }
    ret = (int)callback_ret;
  }
```